File: web4-standard/implementation/sdk/web4/atp.py
"""
Web4 ATP/ADP Value Cycle

Canonical implementation per web4-standard/core-spec/atp-adp-cycle.md.

ATP (Allocation Transfer Packet) / ADP (Allocation Discharge Packet)
is Web4's bio-inspired value metabolism. Value flows through work,
not accumulation.

Key invariant: ATP conservation — total ATP + fees always equals initial.
Validated against: web4-standard/test-vectors/atp/transfer-operations.json
"""

from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, Dict, List, Optional

__all__ = [
    # Classes
    "ATPAccount", "TransferResult",
    # Functions
    "energy_ratio", "transfer", "sliding_scale",
    "check_conservation", "sybil_cost", "fee_sensitivity",
    # Constants
    "ATP_JSONLD_CONTEXT",
]


# ── JSON-LD Context ──────────────────────────────────────────────

ATP_JSONLD_CONTEXT = "https://web4.io/contexts/atp.jsonld"


# ── ATP Account ──────────────────────────────────────────────────

@dataclass
class ATPAccount:
    """
    An entity's ATP balance with lock support.

    ATP tokens exist in three sub-pools:
    - available: can be transferred or locked
    - locked: reserved for pending operations (commit or rollback)
    - adp: discharged tokens tracking past expenditure

    Invariant: total = available + locked (ADP is separate tracking).
    """

    available: float = 0.0
    locked: float = 0.0
    adp: float = 0.0
    initial_balance: float = 0.0

    def __post_init__(self) -> None:
        if self.initial_balance == 0.0:
            self.initial_balance = self.available

    @property
    def total(self) -> float:
        """Total ATP (available + locked)."""
        return self.available + self.locked

    @property
    def energy_ratio(self) -> float:
        """
        Energy ratio: ATP / (ATP + ADP).

        High ratio = entity is earning/conserving.
        Low ratio = entity is spending.
        Zero/zero defaults to 0.5 (neutral).
        (test vector atp-012, atp-013)
        """
        total = self.total + self.adp
        if total == 0:
            return 0.5
        return self.total / total

    def lock(self, amount: float) -> bool:
        """
        Lock ATP for a pending operation.
        Returns False if insufficient available balance.
        """
        if amount > self.available:
            return False
        self.available -= amount
        self.locked += amount
        return True

    def commit(self, amount: float) -> float:
        """
        Commit locked ATP (discharge it).
        Returns actual amount committed (may be less if lock was partial).
        """
        actual = min(amount, self.locked)
        self.locked -= actual
        self.adp += actual
        return actual

    def rollback(self, amount: float) -> float:
        """
        Rollback locked ATP (return to available).
        Returns actual amount rolled back.
        """
        actual = min(amount, self.locked)
        self.locked -= actual
        self.available += actual
        return actual

    def recharge(self, rate: float = 0.1, max_multiplier: float = 3.0) -> float:
        """
        Recharge ATP from ADP pool (test vectors atp-009, atp-010).

        recharge_amount = min(initial * rate, max_balance - current)
        """
        max_balance = self.initial_balance * max_multiplier
        raw_recharge = self.initial_balance * rate
        space = max(0.0, max_balance - self.total)
        actual = min(raw_recharge, space)
        self.available += actual
        return actual

    def to_jsonld(self) -> Dict[str, Any]:
        """
        Serialize to JSON-LD per atp-adp-cycle spec.

        Produces the canonical ATPAccount document structure with:
        - @context header for JSON-LD processors
        - Sub-pool breakdown (available, locked, adp)
        - Computed properties (total, energy_ratio)
        """
        return {
            "@context": [ATP_JSONLD_CONTEXT],
            "@type": "ATPAccount",
            "available": self.available,
            "locked": self.locked,
            "adp": self.adp,
            "initial_balance": self.initial_balance,
            "total": self.total,
            "energy_ratio": self.energy_ratio,
        }

    def to_jsonld_string(self, indent: int = 2) -> str:
        """Serialize to JSON-LD string."""
        return json.dumps(self.to_jsonld(), indent=indent)

    @classmethod
    def from_jsonld(cls, doc: Dict[str, Any]) -> ATPAccount:
        """
        Deserialize from JSON-LD document.

        Accepts both spec JSON-LD format and plain dict format.
        Ignores @context, @type, and computed properties (total, energy_ratio).
        """
        return cls(
            available=doc.get("available", 0.0),
            locked=doc.get("locked", 0.0),
            adp=doc.get("adp", 0.0),
            initial_balance=doc.get("initial_balance", 0.0),
        )

    @classmethod
    def from_jsonld_string(cls, s: str) -> ATPAccount:
        """Deserialize from JSON-LD string."""
        return cls.from_jsonld(json.loads(s))


# ── Transfer Operations ──────────────────────────────────────────

@dataclass
class TransferResult:
    """Result of an ATP transfer."""
    fee: float
    sender_balance: float
    receiver_balance: float
    actual_credit: float
    overflow: float = 0.0

    def to_jsonld(self) -> Dict[str, Any]:
        """
        Serialize to JSON-LD per atp-adp-cycle spec.

        Captures the outcome of an ATP transfer: fee charged,
        resulting balances, actual credit applied, and any overflow.
        """
        doc: Dict[str, Any] = {
            "@context": [ATP_JSONLD_CONTEXT],
            "@type": "TransferResult",
            "fee": self.fee,
            "sender_balance": self.sender_balance,
            "receiver_balance": self.receiver_balance,
            "actual_credit": self.actual_credit,
        }
        if self.overflow > 0:
            doc["overflow"] = self.overflow
        return doc

    def to_jsonld_string(self, indent: int = 2) -> str:
        """Serialize to JSON-LD string."""
        return json.dumps(self.to_jsonld(), indent=indent)

    @classmethod
    def from_jsonld(cls, doc: Dict[str, Any]) -> TransferResult:
        """
        Deserialize from JSON-LD document.

        Accepts both spec JSON-LD format and plain dict format.
        Ignores @context and @type fields.
        """
        return cls(
            fee=doc["fee"],
            sender_balance=doc["sender_balance"],
            receiver_balance=doc["receiver_balance"],
            actual_credit=doc["actual_credit"],
            overflow=doc.get("overflow", 0.0),
        )

    @classmethod
    def from_jsonld_string(cls, s: str) -> TransferResult:
        """Deserialize from JSON-LD string."""
        return cls.from_jsonld(json.loads(s))


def transfer(
    sender: ATPAccount,
    receiver: ATPAccount,
    amount: float,
    fee_rate: float = 0.05,
    max_balance: Optional[float] = None,
) -> TransferResult:
    """
    Transfer ATP between accounts (test vectors atp-001 through atp-003).

    Fee is charged to sender ON TOP of the transfer amount.
    If max_balance is set, overflow is returned to sender.

    Invariant: sender_deducted = amount + fee - overflow
    """
    fee = amount * fee_rate
    total_deduction = amount + fee

    if total_deduction > sender.available:
        raise ValueError(
            f"Insufficient balance: need {total_deduction}, have {sender.available}"
        )

    # Calculate actual credit considering cap
    if max_balance is not None:
        space = max(0.0, max_balance - receiver.available)
        actual_credit = min(amount, space)
        overflow = amount - actual_credit
    else:
        actual_credit = amount
        overflow = 0.0

    # Execute
    sender.available -= total_deduction
    sender.available += overflow  # Return overflow
    receiver.available += actual_credit

    return TransferResult(
        fee=fee,
        sender_balance=sender.available,
        receiver_balance=receiver.available,
        actual_credit=actual_credit,
        overflow=overflow,
    )


# ── Sliding Scale Payment ────────────────────────────────────────

def sliding_scale(
    quality: float,
    base_payment: float,
    zero_threshold: float = 0.3,
    full_threshold: float = 0.7,
) -> float:
    """
    Quality-based sliding scale payment (test vectors atp-005 through atp-007, atp-015).

    Below zero_threshold: 0 payment.
    Between thresholds: linear ramp.
    At or above full_threshold: full base_payment.
    """
    if quality < zero_threshold:
        return 0.0
    if quality >= full_threshold:
        return base_payment
    scale = (quality - zero_threshold) / (full_threshold - zero_threshold)
    return base_payment * scale


# ── Conservation Check ───────────────────────────────────────────

def check_conservation(
    initial_balances: List[float],
    final_balances: List[float],
    total_fees: float,
    tolerance: float = 0.0001,
) -> bool:
    """
    Verify ATP conservation invariant (test vector atp-004).

    initial_total == final_total + total_fees
    """
    initial_total = sum(initial_balances)
    final_total = sum(final_balances)
    return abs(initial_total - (final_total + total_fees)) < tolerance


# ── Energy Ratio (standalone) ────────────────────────────────────

def energy_ratio(atp: float, adp: float) -> float:
    """
    Energy ratio calculation (test vectors atp-012, atp-013).

    atp / (atp + adp), with 0/0 → 0.5 (neutral).
    """
    total = atp + adp
    if total == 0:
        return 0.5
    return atp / total


# ── Sybil Cost Analysis ─────────────────────────────────────────

def sybil_cost(
    num_identities: int,
    hardware_cost: float,
    atp_stake: float,
    fee_rate: float = 0.05,
) -> dict[str, float]:
    """
    Sybil attack cost analysis (test vector atp-011).

    Returns setup cost and per-cycle circular flow loss.
    """
    per_identity = hardware_cost + atp_stake
    total_setup = num_identities * per_identity
    circular_loss = num_identities * atp_stake * fee_rate

    return {
        "total_setup_cost": total_setup,
        "per_identity_cost": per_identity,
        "circular_flow_loss_per_cycle": circular_loss,
    }


# ── Fee Sensitivity ──────────────────────────────────────────────

def fee_sensitivity(amount: float, fee_rates: List[float]) -> List[dict[str, float]]:
    """
    Fee sensitivity sweep (test vector atp-014).

    For each fee rate, compute fee, net received, and total sender cost.
    Fee is additive to sender, not deducted from amount.
    """
    results = []
    for rate in fee_rates:
        fee = amount * rate
        results.append({
            "fee_rate": rate,
            "fee": fee,
            "net_received": amount,  # receiver gets full amount
            "total_sender_cost": amount + fee,
        })
    return results

File: web4-core/src/atp.rs
// Copyright (c) 2026 MetaLINXX Inc.
// SPDX-License-Identifier: AGPL-3.0-or-later

//! ATP/ADP — Allocation Transfer / Discharge Packets
//!
//! Bio-inspired energy metabolism for Web4 societies. ATP is a unit of
//! account (NOT currency) — each society reifies its own resources
//! (compute, attention, hardware, time) into ATP at policies it chooses.
//!
//! Key invariants:
//! - Conservation: sum(initial) == sum(final) + total_fees
//! - Two-state only: tokens exist as ATP or ADP, never both
//! - Pool-managed: exists in society pools, not per-entity wallets
//!
//! Reference: `web4-standard/core-spec/atp-adp-cycle.md`

use crate::error::{Result, Web4Error};
use serde::{Deserialize, Serialize};

/// ATP account — tracks available, locked (escrowed), and discharged tokens.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ATPAccount {
    /// Tokens available for transfer or locking
    pub available: f64,
    /// Tokens locked (escrowed for pending operations)
    pub locked: f64,
    /// Discharged tokens (spent — ADP)
    pub adp: f64,
    /// Initial balance at creation (for recharge calculations)
    pub initial_balance: f64,
}

impl ATPAccount {
    pub fn new(initial: f64) -> Self {
        Self {
            available: initial,
            locked: 0.0,
            adp: 0.0,
            initial_balance: initial,
        }
    }

    /// Total active ATP (available + locked). ADP is separate.
    pub fn total(&self) -> f64 {
        self.available + self.locked
    }

    /// Energy ratio: ATP / (ATP + ADP). High = earning, low = spending.
    /// Returns 0.5 (neutral) if both are zero.
    pub fn energy_ratio(&self) -> f64 {
        let total = self.total() + self.adp;
        if total == 0.0 {
            0.5
        } else {
            self.total() / total
        }
    }

    /// Lock tokens from available → locked (escrow for pending operation).
    pub fn lock(&mut self, amount: f64) -> Result<()> {
        if amount < 0.0 {
            return Err(Web4Error::InvalidInput("Lock amount must be non-negative".into()));
        }
        if self.available < amount {
            return Err(Web4Error::InvalidInput(format!(
                "Insufficient available ATP: {} < {}",
                self.available, amount
            )));
        }
        self.available -= amount;
        self.locked += amount;
        Ok(())
    }

    /// Commit locked tokens → ADP (discharge). Called on successful completion.
    pub fn commit(&mut self, amount: f64) -> Result<f64> {
        if amount < 0.0 {
            return Err(Web4Error::InvalidInput("Commit amount must be non-negative".into()));
        }
        let actual = amount.min(self.locked);
        self.locked -= actual;
        self.adp += actual;
        Ok(actual)
    }

    /// Rollback locked tokens → available. Called on failure/cancellation.
    pub fn rollback(&mut self, amount: f64) -> Result<f64> {
        if amount < 0.0 {
            return Err(Web4Error::InvalidInput("Rollback amount must be non-negative".into()));
        }
        let actual = amount.min(self.locked);
        self.locked -= actual;
        self.available += actual;
        Ok(actual)
    }

    /// Recharge: add ATP up to max_multiplier * initial_balance.
    /// Returns actual amount recharged.
    pub fn recharge(&mut self, rate: f64, max_multiplier: f64) -> f64 {
        let max_balance = self.initial_balance * max_multiplier;
        let raw_recharge = self.initial_balance * rate;
        let space = (max_balance - self.total()).max(0.0);
        let actual = raw_recharge.min(space);
        self.available += actual;
        actual
    }
}

impl Default for ATPAccount {
    fn default() -> Self {
        Self::new(100.0)
    }
}

/// Result of an ATP transfer between two accounts.
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct TransferResult {
    /// Fee charged (additive to sender, not deducted from amount)
    pub fee: f64,
    /// Sender's final available balance
    pub sender_balance: f64,
    /// Receiver's final available balance
    pub receiver_balance: f64,
    /// Amount actually credited to receiver (may be < amount if capped)
    pub actual_credit: f64,
    /// Amount returned to sender if receiver hit max_balance cap
    pub overflow: f64,
}

/// Transfer ATP between two accounts.
///
/// Fee is additive to sender (sender pays amount + fee).
/// If max_balance is set, excess beyond receiver's cap overflows back to sender.
///
/// Conservation invariant: sender_deducted == actual_credit + fee + overflow
pub fn transfer(
    sender: &mut ATPAccount,
    receiver: &mut ATPAccount,
    amount: f64,
    fee_rate: f64,
    max_balance: Option<f64>,
) -> Result<TransferResult> {
    if amount < 0.0 {
        return Err(Web4Error::InvalidInput("Transfer amount must be non-negative".into()));
    }

    let fee = amount * fee_rate;
    let total_deduction = amount + fee;

    if sender.available < total_deduction {
        return Err(Web4Error::InvalidInput(format!(
            "Insufficient ATP: {} < {} (amount {} + fee {})",
            sender.available, total_deduction, amount, fee
        )));
    }

    let (actual_credit, overflow) = if let Some(max) = max_balance {
        let space = (max - receiver.available).max(0.0);
        let credit = amount.min(space);
        (credit, amount - credit)
    } else {
        (amount, 0.0)
    };

    sender.available -= total_deduction;
    sender.available += overflow; // return overflow
    receiver.available += actual_credit;

    Ok(TransferResult {
        fee,
        sender_balance: sender.available,
        receiver_balance: receiver.available,
        actual_credit,
        overflow,
    })
}

/// Sliding scale payment based on quality score.
///
/// Below zero_threshold: pays 0. Above full_threshold: pays full base_payment.
/// Between: linear interpolation.
pub fn sliding_scale(
    quality: f64,
    base_payment: f64,
    zero_threshold: f64,
    full_threshold: f64,
) -> f64 {
    if quality < zero_threshold {
        0.0
    } else if quality >= full_threshold {
        base_payment
    } else {
        base_payment * (quality - zero_threshold) / (full_threshold - zero_threshold)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_account_lifecycle() {
        let mut account = ATPAccount::new(100.0);
        assert_eq!(account.total(), 100.0);
        assert_eq!(account.energy_ratio(), 1.0); // all ATP, no ADP

        // Lock 30
        account.lock(30.0).unwrap();
        assert_eq!(account.available, 70.0);
        assert_eq!(account.locked, 30.0);
        assert_eq!(account.total(), 100.0);

        // Commit 20 (discharge to ADP)
        account.commit(20.0).unwrap();
        assert_eq!(account.locked, 10.0);
        assert_eq!(account.adp, 20.0);
        assert_eq!(account.total(), 80.0);

        // Energy ratio: 80 / (80 + 20) = 0.8
        assert!((account.energy_ratio() - 0.8).abs() < 1e-10);

        // Rollback remaining locked
        account.rollback(10.0).unwrap();
        assert_eq!(account.available, 80.0);
        assert_eq!(account.locked, 0.0);
    }

    #[test]
    fn test_transfer_conservation() {
        let mut sender = ATPAccount::new(100.0);
        let mut receiver = ATPAccount::new(50.0);

        let result = transfer(&mut sender, &mut receiver, 30.0, 0.05, None).unwrap();

        assert_eq!(result.fee, 1.5);
        assert_eq!(result.actual_credit, 30.0);
        assert_eq!(result.overflow, 0.0);
        assert_eq!(sender.available, 68.5); // 100 - 30 - 1.5
        assert_eq!(receiver.available, 80.0); // 50 + 30

        // Conservation: sender lost 31.5, receiver gained 30, fee = 1.5
        // 31.5 == 30 + 1.5 ✓
    }

    #[test]
    fn test_transfer_with_max_balance() {
        let mut sender = ATPAccount::new(100.0);
        let mut receiver = ATPAccount::new(90.0);

        // Receiver can only take 10 more (max_balance = 100)
        let result = transfer(&mut sender, &mut receiver, 30.0, 0.0, Some(100.0)).unwrap();

        assert_eq!(result.actual_credit, 10.0);
        assert_eq!(result.overflow, 20.0);
        assert_eq!(sender.available, 90.0); // 100 - 30 + 20 overflow
        assert_eq!(receiver.available, 100.0);
    }

    #[test]
    fn test_recharge() {
        let mut account = ATPAccount::new(100.0);
        account.available = 50.0; // spent some

        let recharged = account.recharge(0.1, 3.0);
        assert_eq!(recharged, 10.0); // 100 * 0.1 = 10, space = 300 - 50 = 250
        assert_eq!(account.available, 60.0);
    }

    #[test]
    fn test_sliding_scale() {
        assert_eq!(sliding_scale(0.1, 100.0, 0.3, 0.7), 0.0);
        assert!((sliding_scale(0.5, 100.0, 0.3, 0.7) - 50.0).abs() < 1e-10);
        assert_eq!(sliding_scale(0.8, 100.0, 0.3, 0.7), 100.0);
    }

    #[test]
    fn test_zero_balance_energy_ratio() {
        let account = ATPAccount::new(0.0);
        assert_eq!(account.energy_ratio(), 0.5); // neutral
    }

    #[test]
    fn test_insufficient_balance() {
        let mut sender = ATPAccount::new(10.0);
        let mut receiver = ATPAccount::new(0.0);
        let result = transfer(&mut sender, &mut receiver, 20.0, 0.0, None);
        assert!(result.is_err());
    }
}
